![logo_ironhack_blue 7](https://user-images.githubusercontent.com/23629340/40541063-a07a0a8a-601a-11e8-91b5-2f13e4e6b441.png)

# Lab | Feature engineering

- In this lab, you will use `learningSet.csv` file which you have already cloned in the previous activities. 

In [2]:
# 📚 Basic libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import warnings

In [3]:
# ⚙️ Settings
pd.set_option('display.max_columns', None) # display all columns
warnings.filterwarnings('ignore') # ignore warnings

In [4]:
data = pd.read_csv('learningSet.csv')

- Check for null values in the numerical columns.

In [5]:
num = data.select_dtypes(include=[np.number])
num.isna().sum()

ODATEDW         0
TCODE           0
DOB             0
AGE         23665
NUMCHLD     83026
            ...  
TARGET_B        0
TARGET_D        0
HPHONE_D        0
RFA_2F          0
CLUSTER2      132
Length: 407, dtype: int64

In [6]:
# Crazy length. We will focus on the columns that the lab wants us to clean:

- Use appropriate methods to clean the columns `GEOCODE2`, `WEALTH1`, `ADI`, `DMA`, and `MSA`.
- Use appropriate EDA technique where ever necessary.

### Cleaning GEOCODE2

In [7]:
data['GEOCODE2'].value_counts()

GEOCODE2
A    34484
B    28505
D    16580
C    15524
       187
Name: count, dtype: int64

In [8]:
# So, there are 4 discrete categories with 187 NaN, let's visualize it better adding empty spaces
empty_geocode = data['GEOCODE2'].eq(' ').sum()
nan_geocode = data['GEOCODE2'].isna().sum() # Thanks again ~Supa to share this code in Discord !
geocode = pd.DataFrame({"Count": [empty_geocode, nan_geocode]}, index=["Empty_Spaces", "NaN_Values"]) # Dictionaries FTW !
geocode

Count
Empty_Spaces    187
NaN_Values      132

In [9]:
# So, for GEOCODE we have 187 empty spaces an 132 NaN
geo_mode = data['GEOCODE2'].mode()[0] # Selecting geocode mode
geo_mode

'A'

In [10]:
its_significant = empty_geocode + nan_geocode
result = round(its_significant*100/len(data['GEOCODE2']), 2)
result

0.33

In [11]:
# It seems to be not very significant. So we will proceed to replace all NaN and Empty Spaces with the majority class.
# Important note --> We are doing this without knowing exactly the meaning of each columnn
data['GEOCODE2'] = data['GEOCODE2'].replace(' ', geo_mode).fillna(geo_mode)
# And again... we check !
empty_geocode = data['GEOCODE2'].eq(' ').sum()
nan_geocode = data['GEOCODE2'].isna().sum() # Thanks again ~Supa to share this code in Discord !
geocode = pd.DataFrame({"Count": [empty_geocode, nan_geocode]}, index=["Empty_Spaces", "NaN_Values"]) # Dictionaries FTW !
geocode

Count
Empty_Spaces      0
NaN_Values        0

### Cleaning WEALTH1

In [12]:
data['WEALTH1'].value_counts()

WEALTH1
9.0    7585
8.0    6793
7.0    6198
6.0    5825
5.0    5280
4.0    4810
3.0    4237
2.0    4085
1.0    3454
0.0    2413
Name: count, dtype: int64

In [13]:
empty_wealth = data['WEALTH1'].eq(' ').sum()
nan_wealth = data['WEALTH1'].isna().sum() # Thanks again ~Supa to share this code in Discord !
wealth = pd.DataFrame({"Count": [empty_wealth, nan_wealth]}, index=["Empty_Spaces", "NaN_Values"]) # Dictionaries FTW !
wealth

Count
Empty_Spaces      0
NaN_Values    44732

In [14]:
its_significant = empty_wealth + nan_wealth
result = round(its_significant*100/len(data['WEALTH1']), 2)
result

46.88

In [15]:
# Difficult to decide without knowing the exact meaning of the column, but wealth should be an important one. We check the dataset for other wealth columns
data.head(0)

Empty DataFrame
Columns: [ODATEDW, OSOURCE, TCODE, STATE, ZIP, MAILCODE, PVASTATE, DOB, NOEXCH, RECINHSE, RECP3, RECPGVG, RECSWEEP, MDMAUD, DOMAIN, CLUSTER, AGE, AGEFLAG, HOMEOWNR, CHILD03, CHILD07, CHILD12, CHILD18, NUMCHLD, INCOME, GENDER, WEALTH1, HIT, MBCRAFT, MBGARDEN, MBBOOKS, MBCOLECT, MAGFAML, MAGFEM, MAGMALE, PUBGARDN, PUBCULIN, PUBHLTH, PUBDOITY, PUBNEWFN, PUBPHOTO, PUBOPP, DATASRCE, MALEMILI, MALEVET, VIETVETS, WWIIVETS, LOCALGOV, STATEGOV, FEDGOV, SOLP3, SOLIH, MAJOR, WEALTH2, GEOCODE, COLLECT1, VETERANS, BIBLE, CATLG, HOMEE, PETS, CDPLAY, STEREO, PCOWNERS, PHOTO, CRAFTS, FISHER, GARDENIN, BOATS, WALKER, KIDSTUFF, CARDS, PLATES, LIFESRC, PEPSTRFL, POP901, POP902, POP903, POP90C1, POP90C2, POP90C3, POP90C4, POP90C5, ETH1, ETH2, ETH3, ETH4, ETH5, ETH6, ETH7, ETH8, ETH9, ETH10, ETH11, ETH12, ETH13, ETH14, ETH15, ETH16, AGE901, ...]
Index: []

In [16]:
data['WEALTH2'].value_counts()

WEALTH2
9.0    6523
8.0    5975
7.0    5684
6.0    5497
5.0    5351
4.0    5074
3.0    5046
2.0    4971
1.0    4191
0.0    3277
Name: count, dtype: int64

In [17]:
empty_wealth2 = data['WEALTH2'].eq(' ').sum()
nan_wealth2 = data['WEALTH2'].isna().sum() # Thanks again ~Supa to share this code in Discord !
wealth2 = pd.DataFrame({"Count": [empty_wealth, nan_wealth]}, index=["Empty_Spaces", "NaN_Values"]) # Dictionaries FTW !
wealth2

Count
Empty_Spaces      0
NaN_Values    44732

In [18]:
its_significant2 = empty_wealth2 + nan_wealth2
result2 = round(its_significant2*100/len(data['WEALTH2']), 2)
result2

45.93

- So, we have two columns named WEALTH, WEALTH1 having 46.88% of NaN and WEALTH2 45.93%
-  Also, we don't know the meaning of the columns.

In this case... we can:

1. Drop the column. It really has TOO MANY `NaN` values, but dropping the column without further consideration could not be considered best practices.
2. Fill it with 0, to represent unknown wealth values per donator. But, this can lead to skewed data, if `WEALTH1` and `WEALTH2` have other meanings or have a big impact on predicting the target.
3. Predicting the wealth values with linear regression.

In [19]:
# Drop the whole column. Almost half of the data is NaN, and it dosen't seem to be correlated with the targets:
corr_target_D = data['TARGET_D'].corr(data['WEALTH1'])
corr_target_B = data['TARGET_B'].corr(data['WEALTH1'])

print(f"Correlation of WEALTH1 with TARGET_D: {corr_target_D}")
print(f"Correlation of WEALTH1 with TARGET_B: {corr_target_B}")

Correlation of WEALTH1 with TARGET_D: 0.024867311730976026
Correlation of WEALTH1 with TARGET_B: 0.011248763279752506


It dosen't seem to be correlated. After some consideration, we drop the column.

If, in future labs/projects, we need WEALTH1 to create a model to predict `TARGET_D` and `TARGET_B` values, we can fine-tune our model predicting WEALTH or trying other methods.

In [20]:
data = data.drop('WEALTH1', axis=1)

### Cleaning ADI

In [21]:
data['ADI'].value_counts()

ADI
13.0     7296
51.0     4622
65.0     3765
57.0     2836
105.0    2617
         ... 
651.0       1
103.0       1
601.0       1
161.0       1
147.0       1
Name: count, Length: 204, dtype: int64

In [22]:
empty_adi = data['ADI'].eq(' ').sum()
nan_adi = data['ADI'].isna().sum() # Thanks again ~Supa to share this code in Discord !
adi = pd.DataFrame({"Count": [empty_adi, nan_adi]}, index=["Empty_Spaces", "NaN_Values"]) # Dictionaries FTW !
adi

Count
Empty_Spaces      0
NaN_Values      132

In [23]:
its_significant = empty_adi + nan_adi
result = round(its_significant*100/len(data['ADI']), 2)
result

0.14

In [24]:
adi_mode = data['ADI'].mode()[0]
data['ADI'] = data['ADI'].fillna(adi_mode)
adi = pd.DataFrame({"Count": [empty_adi, nan_adi]}, index=["Empty_Spaces", "NaN_Values"]) # Dictionaries FTW !
adi

Count
Empty_Spaces      0
NaN_Values      132

### Cleaning DMA

In [25]:
data['DMA'].value_counts()

DMA
803.0    7296
602.0    4632
807.0    3765
505.0    2839
819.0    2588
         ... 
569.0       1
554.0       1
584.0       1
552.0       1
516.0       1
Name: count, Length: 206, dtype: int64

In [26]:
empty_dma = data['DMA'].eq(' ').sum()
nan_dma = data['DMA'].isna().sum()
dma = pd.DataFrame({"Count": [empty_dma, nan_dma]}, index=["Empty_Spaces", "NaN_Values"])
dma

Count
Empty_Spaces      0
NaN_Values      132

In [27]:
its_significant = empty_dma + nan_dma
result = round(its_significant*100/len(data['DMA']), 2)
result

0.14

In [28]:
dma_mode = data['DMA'].mode()[0]
data['DMA'] = data['DMA'].fillna(dma_mode)
empty_dma = data['DMA'].eq(' ').sum()
nan_dma = data['DMA'].isna().sum()
dma = pd.DataFrame({"Count": [empty_dma, nan_dma]}, index=["Empty_Spaces", "NaN_Values"])
dma

Count
Empty_Spaces      0
NaN_Values        0

### Cleaning MSA

In [29]:
data['MSA'].value_counts()

MSA
0.0       21333
4480.0     4606
1600.0     4059
2160.0     2586
520.0      1685
          ...  
9140.0        1
3200.0        1
9280.0        1
743.0         1
8480.0        1
Name: count, Length: 298, dtype: int64

In [30]:
empty_msa = data['MSA'].eq(' ').sum()
nan_msa = data['MSA'].isna().sum()
msa = pd.DataFrame({"Count": [empty_msa, nan_msa]}, index=["Empty_Spaces", "NaN_Values"])
msa

Count
Empty_Spaces      0
NaN_Values      132

In [31]:
msa_mode = data['MSA'].mode()[0]
data['MSA'] = data['MSA'].fillna(msa_mode)
empty_msa = data['MSA'].eq(' ').sum()
nan_msa = data['MSA'].isna().sum()
msa = pd.DataFrame({"Count": [empty_msa, nan_msa]}, index=["Empty_Spaces", "NaN_Values"])
msa

Count
Empty_Spaces      0
NaN_Values        0